File: src/listener.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "auv_homework/SensorData.h"

#include <sstream>

// to access the topic on callback function
ros::Publisher counter_pub;
// to keep track of message count
int msg_count = 0;

void sensorDataCallback(const auv_homework::SensorData::ConstPtr &data)
{
  msg_count++;

  std_msgs::String msg;
  std::stringstream ss;
  ss << "message count: " << msg_count;
  msg.data = ss.str();

  counter_pub.publish(msg);

  // log the data read from sensor_data topic, and the string published to message_count topic
  ROS_INFO("I heard sensor-%d published: %f", data->id, data->value);
  ROS_INFO("I published: %s", msg.data.c_str());
}

int main(int argc, char **argv)
{
  // init a ROS node
  ros::init(argc, argv, "listener");
  ros::NodeHandle n;

  // subscribe to sensor_data topic with our callback func
  ros::Subscriber sub = n.subscribe("sensor_data", 10, sensorDataCallback);

  // advertise the message_count topic with the msg_type of String
  counter_pub = n.advertise<std_msgs::String>("message_count", 10);

  ros::spin();
  return 0;
}

File: scripts/listener.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String
from auv_homework.msg import SensorData


message_count = 0
pub = None


def callback(data):
    global message_count
    message_count += 1

    message_count_log = "message_count: %s" % message_count

    pub.publish(message_count_log)

    rospy.loginfo("I heard sensor-%s published: %s", data.id, data.value)
    rospy.loginfo("I published: %s", message_count_log)


def listener():
    global pub

    rospy.init_node("listener", anonymous=True)

    rospy.Subscriber("sensor_data", SensorData, callback)

    pub = rospy.Publisher("message_count", String, queue_size=10)

    rospy.spin()


if __name__ == "__main__":
    listener()
